src/libexpr/value.hh

```diff
@@ -84,152 +64,445 @@ class ExternalValueBase
+
+         "--with-freetype2-library=" + freetype + "/lib"
+
+         where `freetype' is a derivation (or a source to be copied
+         to the store).  If we just concatenated the strings without
+         keeping track of the referenced store paths, then if the
+         string is used as a derivation attribute, the derivation
+         will not have the correct dependencies in its inputDrvs and
+         inputSrcs.
+
+         The semantics of the context is as follows: when a string
+         with context C is used as a derivation attribute, then the
+         derivations in C will be added to the inputDrvs of the
+         derivation, and the other store paths in C will be added to
+         the inputSrcs of the derivations.
+
+         For canonicity, the store paths should be in sorted order. */
+    void setStringNoCopy(const char * str, const char * * context) {
+        setFirst(LFT_1, str);
+        setSecond(LST_String, context);
+        assert(type() == tString);
+    }
+    void setString(const Symbol & s) {
+        setStringNoCopy(((const string &) s).c_str(), nullptr);
+    }
+    void setString(const char * str);
+    void setString(const string & s) {
+        setString(s.c_str());
+    }
+    void setString(const string & s, const PathSet & context);
+    const char * asString() const {
+        assert(type() == tString);
+        return getFirst<const char *>(LFT_1);
+    }
+    const char * * asStringContext() const {
+        assert(type() == tString);
+        return getSecond<const char * *>(LST_String);
+    }
 
+    void setPathNoCopy(const char * path) {
+        setFirstHigh(HFT_Path);
+        setSecond(path);
+        assert(type() == tPath);
+    }
+    void setPath(const char * path);
+    const char * asPath() const {
+        assert(type() == tPath);
+        return getSecond<const char *>();
+    }
 
-static inline void mkApp(Value & v, Value & left, Value & right)
-{
-    v.type = tApp;
-    v.app.left = &left;
-    v.app.right = &right;
-}
+    void setAttrs(Bindings * attrs) {
+        setFirstHigh(HFT_Attrs);
+        setSecond(attrs);
+        assert(type() == tAttrs);
+    }
+    Bindings * asAttrs() const {
+        assert(type() == tAttrs);
+        return getSecond<Bindings *>();
+    }
 
+    void setList() {
+        setFirstHigh(HFT_List0);
+        assert(type() == tList0);
+    }
+    void setList(Value * e0) {
+        setFirstHigh(HFT_List1);
+        setSecond(e0);
+        assert(type() == tList1);
+    }
+    void setList(Value * e0, Value * e1) {
+        setFirst(LFT_0, e0);
+        setSecond(LST_List2, e1);
+        assert(type() == tList2);
+    }
+    void setList(Value * * elems, size_t size) {
+        setFirst(LFT_0, elems);
+        setSecond(LST_ListN, size * LST_MAX);
+        assert(type() == tListN);
+    }
 
-static inline void mkStringNoCopy(Value & v, const char * s)
-{
-    v.type = tString;
-    v.string.s = s;
-    v.string.context = 0;
-}
+    class asList {
```

Comment:
Why is `asList` and `asMutableList` needed? Seems kind of inefficient to materialize the list contents like this, though it's probably mostly optimized away.